add strategy paper for observability (#695)
kyma-project/community: concepts/observability-strategy/strategy.md

- Very limited integration possibilities to external systems. Integration is usually needed for different reasons, such as cross-cluster correlation, forensic analysis, or long-term storage. Kyma's integration points are not vendor-neutral.
- Very limited configuration options for data enrichment and filtering. Users want to enrich the data with data relevant for there environments, like cluster names. Furthermore, they want to filter out non-relevant log lines or log attributes within a line to save resources/money in the backend.
- Storage backends are non-scalable, so they can be used only in limited scenarios. Users cannot upgrade the backend into a scalable setup, nor integrate with other solutions.

## Requirements

Integration (and with that, moving the focus away from in-cluster backends) is the key to open up the stack for a broad range of use cases. Users can bring their own backends if they already use a commercial offering or run their own infrastructure. The data can be stored outside the cluster in a managed offering, shared with the data of multiple clusters, away from any tampering or deletion attempt of a hacker, to name just a few advantages.

> Providing ready-to-use in-cluster backends is necessarily opinionated, does not cover all usage scenarios, and does not fit into Kyma's goal of providing the Kubernetes building blocks to integrate into the the SAP ecosystem. Also, the licensing issues (particularly with Grafana and Loki, or ElasticSearch as an alternative backend technology) show that an opinionated stack is problematic. It's better to handle opinionation by integrating with actual managed services. 

Reviewer: ```suggestion
Providing ready-to-use in-cluster backends is necessarily opinionated, does not cover all usage scenarios, and does not fit into Kyma's goal of providing the Kubernetes building blocks to integrate into the the SAP ecosystem. Also, the licensing issues (particularly with Grafana and Loki, or Elasticsearch as an alternative backend technology) show that an opinionated stack is problematic. It's better to handle opinionation by integrating with actual managed services. 
```

2. **Transform tracing component into a blueprint**: The tracing component bundles a Jaeger-all-in-one installation pre-integrated with Kyma's Istio, Serverless, and Eventing components. The bundle has a very lightweight setup that is mainly for demo purposes. It should be turned into a blueprint to save maintenance efforts. In the same step, Kyma must be adjusted to leverage Istio's telemetry API so that the telemetry can be activated at runtime.

> 2. [**Configurable Logging**](https://github.com/kyma-project/kyma/issues/11236): Introduce a configurable log collector and support log pipeline configurations at runtime with focus on the SAP ecosystem as MVP. Accessing application logs is the first and most common way of troubleshooting applications and is the minimal feature that users expect.

Reviewer: It's rendered correctly as steps 1-7, but in the raw view it's a bit confusing we've got two step 2s

3. Signal Delivery: The signals need to be consumed by some party, either running inside or outside the cluster. It can be an aggregating layer or the backend directly. In any way, the shipment of the data needs to be configurable and should be based on a neutral protocol so that any aggregation layer or backend can be integrated.

The optional parts where Kyma can provide features are for the stages to the right (aggregation/storage/analysis). Here, usually a oppinionated solution must be picked as there is no common technology available. Such oppinionated solution can be pre-configured to provide either a starting point or a full-blown solution. However, these aspects will usually not cover all usage and operations/scalability scenarios. Furthermore, it will require a huge investment in order to provide just another backend solution.

> With that, the goal of the kyma observability should be the seemless enablement of the stages on the left, allowing the ingestion of the signals into any backend system, opening up plenty usage scenarios. Kyma should provide blueprints for the stages on the right, especially focussed on the SAP ecosystem enablement, in order to have at least some guidance for these aspects.

Reviewer: ```suggestion

### Goals: Summary

The goal of the Kyma observability stack should be the seamless enablement of the stages on the left, allowing the ingestion of the signals into any backend system, opening up plenty usage scenarios. Kyma should provide at least blueprints as guidance for the stages on the right, especially for enabling the SAP ecosystem.
```

To sum it up, the goals of kyma observability should be:

> 1. provide guides on how to do instrumentation (with potential helpers and auto-instrumentation options)

Reviewer: ```suggestion
1. Provide guides on instrumentation (with potential helpers and auto-instrumentation options).
```

> 2. collect resulting signals instantly when guides are followed, provide customization options for border cases

Reviewer: ```suggestion
2. Collect resulting signals instantly when guides are followed, provide customization options for special cases.
```

![b](./assets/future.drawio.svg)

As mentioned before, the technology and protocols for the signal collection are different for the signal types. Logs are getting tailed from container log files, metrics are getting pulled via the prometheus format usually and traces are getting pushed via OTLP. With that, also the pre-integration so that typical signals are getting collected instantly are different per type.
That's why the concrete concepts for the different types are different and will be discussed in more detail in the individual sections:
* [Concept - Configurable Logging](./configurable-logging/README.md) 
* [Concept - Configurable Monitoring](./configurable-monitoring/README.md)
* [Concept - Configurable Tracing](./configurable-tracing/README.md)

> All concepts will follow general rules and will provide harmonized user APIs:

Reviewer: ```suggestion
All concepts follow general rules and will provide harmonized user APIs:
```

> - **By default signals of user workload will be collected**: When a workload gets instrumented as outlined in the best practice, the telemetry data will be instantly available in a pipeline without further configuration, just define a pipeline with an output and it works

Reviewer: ```suggestion
- **By default, signals of user workloads are collected**: If a workload is instrumented as outlined in the best practice guide, the telemetry data is instantly available in a pipeline without further configuration. Just define a pipeline with an output and it works.
```

What's the difference between "guideline", "best practice (guide)", and "blueprint"? Are they synonyms for us or do we differentiate here? If they're used as synonyms, we should settle for just one term and drop the others to improve consistency and simplicity.

> - **One input protocol per type**: Only one common input protocol (like OTLP) will be supported per signal type having best practices on how to integrate with.

Reviewer: ```suggestion
- **One input protocol per type**: Only one common input protocol (like OTLP) is supported per signal type. Kyma provides best practice guides on how to integrate with it.
```

> - **Custom input protocols can be integrated**: When a workload is not using the supported protocol for signal exposure, there is a way to have custom transformation to the supported protocol in place, usually related with additional operational effort (like running a dedicated otel-collector sidecar/deployment in the users namespace)

Reviewer: ```suggestion
- **Custom input protocols can be integrated**: When a workload doesn't use the supported protocol for signal exposure, users can have custom transformation to the supported protocol, usually with additional operational effort (like running a dedicated Otel collector sidecar or deployment in the user's Namespace)
```

- **Harmonized API**: Signal pipelines are defined via well-defined Kubernetes CRDs only belonging to one group and following similar semantics.

> - **One output protocol per type**: Only one common output protocol (like OTLP) will be supported per signal type.

Reviewer: ```suggestion
- **One output protocol per type**: Only one common output protocol (like OTLP) is supported per signal type.
```

> - **Custom output protocols can be integrated**: there is a way to have custom transformation from the supported protocol in place, usually related with additional operational effort (like running a dedicated otel-collector sidecar/deployment in the users namespace). Furthermore, typical output plugins of the used collector technology will be offered as part of a pipeline definition, however the plugins will be offered with limited support only.

Reviewer: ```suggestion
- **Custom output protocols can be integrated**: Users can have custom transformation from the supported protocol, usually with additional operational effort (like running a dedicated Otel Collector sidecar or deployment in the users' Namespace). Furthermore, Kyma offers typical output plugins of the used collector technology as part of a pipeline definition; however, only with limited support.
```

> - **Blueprints are fully integrated**: Blueprints for local backend deployments will be integrated with the used pipeline mechanism so that they work instantly

Reviewer: ```suggestion
- **Blueprints are fully integrated**: Blueprints for local backend deployments are integrated with the used pipeline mechanism, so that they work instantly.
```

2. **Transform tracing component into a blueprint**: The tracing component bundles a Jaeger-all-in-one installation pre-integrated with Kyma's istio, serverless and eventing. The bundle is providing such lightweight setup, that it is mainly for demo purposes and should be consequently turned into a blueprint to save maintenance efforts. While turning it into a blueprint, kyma should be adjusted to leverage Istio's telemetry API to have the telemetry activatable at runtime.

> 2. [**Configurable Logging**](https://github.com/kyma-project/kyma/issues/11236): Introduce a configurable log collector and support log pipeline configurations at runtime with focus on the SAP eco-system as MVP. Accessing application logs is the first and most common way of troubleshooting applications and is the minimal feature users will expect.

Reviewer: ```suggestion
2. [**Configurable Logging**](https://github.com/kyma-project/kyma/issues/11236): Introduce a configurable log collector and support log pipeline configurations at runtime with focus on the SAP ecosystem as MVP. Accessing application logs is the first and most common way of troubleshooting applications and is the minimal feature that users expect.
```

> 3. **Transform logging component into a blueprint**: As soon as the application logs can be out-streamed into external systems, there is an alternative available for the in-cluster logging solution. The Loki stack (Loki + Grafana) can be turned into a blueprint having the advantage of using latest Loki versions (license problem). Again, the upstream loki helm chart can be used with a provided values.yaml with instant integration.

Reviewer: ```suggestion
3. **Transform logging component into a blueprint**: As soon as the application logs can be out-streamed to external systems, Kyma can choose alternatives for the in-cluster logging solution. The Loki stack (Loki and Grafana) can be turned into a blueprint with the advantage of using the latest Loki versions (solving the license problem). Again, users get instant integration with the upstream Loki Helm chart and a provided `values.yaml` file.
```

### Optional feature set

Optionally, Kyma can provide features for the stages to the right (aggregation, storage, and analysis). Here, users usually must pick an opinionated solution because there is no common technology available. Kyma can support such opinionated solutions by pre-configuring either a starting point or a full-blown solution. However, these aspects will usually not cover all usage, operations, and scalability scenarios. Furthermore, huge investments are necessary to provide any additional backend solution.

### Goals: Summary

The goal of the Kyma observability stack should be the seamless enablement of the stages on the left, allowing the ingestion of the signals into any backend system, opening up plenty usage scenarios. Kyma should provide at least blueprints as guidance for the stages on the right, especially for enabling the SAP ecosystem.
To sum it up, the goals of Kyma observability should be:

> 1. Provide guides on instrumentation (with potential helpers and auto-instrumentation options).

Author: It is more an order by stages, no priority

> 1. Provide guides on instrumentation (with potential helpers and auto-instrumentation options).

Reviewer: Then I'd maybe use a bullet list, not a numbered list.

With such limited integration possibilities and providing only a very lightweight in-cluster solution, the current observability stack is missing a lot of usage scenarios, and the potential audience is very narrow. Introducing vendor-neutral integration possibilities and moving the focus away from backend solutions will enable The Kyma runtime for a broad audience and enable adoption.

## Current Situation 

Observability can be split up in the following stages:
1. Instrumentation of the user application
2. Collection and preprocessing of the signals from the user application and the surrounding infrastructure, including metadata enrichment
3. Delivery of the signal to an backend for storage
4. Optionally, aggregation of the signals
5. Storage of the signals
6. Analysis, querying, and dashboarding of the signals

![a](./assets/strategy-stages.drawio.svg)

> The current Kyma observability stack is spreading over all stages and is providing a lightweight end-to-end setup for basic needs. Hereby, configuration possibilities are available at installation time mainly.

Reviewer: ```suggestion
The current Kyma observability stack covers all stages, providing a lightweight end-to-end setup for basic needs. It can mainly be configured at installation.
```

![Current observability stack](./assets/strategy-current.drawio.svg "Current observability stack")

- Application Logs
  - Instrumentation is done by printing logs to stdout/stderr as recommended by Kubernetes best practices, system components are following that guide and logs are available already; no other way of getting logs collected
  - FluentBit as log collector defines a hardcoded pipeline for collecting, enriching and pushing the logs to a backend; no further configuration of pipelines at runtime, especially additional outputs for external systems cannot be configured
  - Storage is realized by a lightweight (non-scalable) Loki installation, no configuration at runtime
  - Reporting is implemented by the log explorer of Grafana

> - Metrics

Reviewer: Consider formatting this as h3 and reducing the indentation for the sub-bullets.

```suggestion
### Metrics
```

  - For instrumentation workload needs to expose metrics in the prometheus-compatible format, system components are already doing that; no other ways of exporting metrics is supported like using the OTLP push based protocol
  - Collection of metrics from system and custom workloads is done by a lightweight prometheus installation; configuration of collection can be defined at runtime but the storage might not scale in accordingly requiring adjustments at deploy time; configuration of outputs is not possible at runtime and only prometheus-specific protocols are supported (forward or federation)
  - Prometheus is collecting and storing the metrics; setup is non-scalable and resource settings cannot be configured at runtime
  - Reporting is done via a Grafana installation which loads pre-bundled dashboards; dashboards can be added at runtime

> - Traces

Reviewer: Consider formatting this as h3 and reducing the indentation for the sub-bullets.

```suggestion
### Traces
```

  - 

> The current solution provides a feature-rich end-to-end setup at a first glance. However, at a second look users realize soon the major drawbacks and usually will require additional stacks deployed besides:

Reviewer: ```suggestion
### Drawbacks

At a first glance, the current solution provides a feature-rich end-to-end setup. However, at a second look, users notice major drawbacks and usually need additional stacks.
```

> - Very limited integration possibilities into external systems; that is usually required for different reasons (cross-cluster correlation, forensic analysis, long time storage,..); integration points are not vendor-neutral

Reviewer: ```suggestion
- Very limited integration possibilities to external systems. Integration is usually needed for different reasons, such as cross-cluster correlation, forensic analysis, or long-term storage. Kyma's integration points are not vendor-neutral.
```

> - Very limited configuration options for data enrichment/filtering; users are interested in specific aspects for specific scenarios

Reviewer: ```suggestion
- Very limited configuration options for data enrichment and filtering. Users are interested in specific aspects for specific scenarios.
```
The second sentence has extremely little information value. You can do better.

> - Storage backends are non-scalable and with can be used only in limited scenarios. There is no option to upgrade the backend into a scalable setup but also no way to integrate with other solutions

Reviewer: ```suggestion
- Storage backends are non-scalable, so they can be used only in limited scenarios. Users cannot upgrade the backend into a scalable setup, nor integrate with other solutions.
```

> ### Requirements

Reviewer: I think this should be h2 not h3, otherwise it's part of the "Current situation" section.

```suggestion
## Requirements
```

Integration (and with that, moving the focus away from in-cluster backends) is the key to open up the stack for a broad range of use cases. Users can bring their own backends if they already use a commercial offering or run their own infrastructure. The data can be stored outside the cluster in a managed offering, shared with the data of multiple clusters, away from any tampering or deletion attempt of a hacker, to name just a few advantages.

Providing ready-to-use in-cluster backends is necessarily opinionated, does not cover all usage scenarios, and does not fit into Kyma's goal of providing the Kubernetes building blocks to integrate into the the SAP ecosystem. Also, the licensing issues (particularly with Grafana and Loki, or ElasticSearch as an alternative backend technology) show that an opinionated stack is problematic. It's better to handle opinionation by integrating with actual managed services. 

> In order to support integration as a key element, it is mandatory that a Kyma runtime supports the stages that must happen within a runtime itself (instrumentation, collection, delivery). Thus, the following points will be the major goal of the Kyma feature set:

Reviewer: Why don't you like my headline suggestion?

```suggestion
### Mandatory features

To support integration as a key element, a Kyma runtime must support the stages that happen within the runtime itself (instrumentation, collection, delivery). Thus, the following points are mandatory for the future Kyma observability stack:
```
Trying to simplify that further.

1. Instrumentation support: Users usually need custom code for the instrumentation of a workload to expose typical signals like logs, traces, and metrics. Kyma can support common aspects of these tasks with guidelines, best practice guides, and ways of auto-instrumentation (like Istio tracing).
2. Signal collection: The runtime must provide the basic infrastructure to collect the emitted signals (like a unified log collector). If the user has followed the provided guides, that infrastructure should collect the signals instantly. Special cases (like specific protocols not backed by guides and infrastructure) should be supported by simple customizations (plug in a custom trace converter). Furthermore, the infrastructure must already enrich the signals of the workloads with metadata of the infrastructure, because this cannot happen at a later stage.
3. Signal delivery: The signals must be consumed by some party, either running inside or outside the cluster. It can be an aggregating layer or the backend directly. In either case, the shipment of the data must be configurable and should be based on a neutral protocol, so that any aggregation layer or backend can be integrated.

> Optionally, Kyma can provide features for the later stages (aggregation, storage, and analysis). Here, users typically must pick an opinionated solution because there is no common technology available. Kyma can support such opinionated solutions by pre-configuring either a starting point or a full-blown solution. However, these aspects will usually not cover all usage, operations, and scalability scenarios. Furthermore, huge investments are necessary to provide any additional backend solution.

Reviewer: Why don't you like my headline suggestion?

```suggestion
### Optional features

Optionally, Kyma can provide features for the later stages (aggregation, storage, and analysis). Here, users typically must pick an opinionated solution because there is no common technology available. Kyma can support such opinionated solutions by pre-configuring either a starting point or a full-blown solution. However, these aspects will usually not cover all usage, operations, and scalability scenarios. Furthermore, huge investments are necessary to provide any additional backend solution.
```

![a](./assets/strategy-stages.drawio.svg)

The current Kyma observability stack covers all stages, providing a lightweight end-to-end setup for basic needs. It can mainly be configured at installation.

![Current observability stack](./assets/strategy-current.drawio.svg "Current observability stack")

### Application Logs
  - Instrumentation is done by printing logs to stdout/stderr as recommended by Kubernetes best practices, system components are following that guide and logs are available already; no other way of getting logs collected
  - FluentBit as log collector defines a hardcoded pipeline for collecting, enriching and pushing the logs to a backend; no further configuration of pipelines at runtime, especially additional outputs for external systems cannot be configured
  - Storage is realized by a lightweight (non-scalable) Loki installation, no configuration at runtime
  - Reporting is implemented by the log explorer of Grafana
### Metrics

>   - For instrumentation workload needs to expose metrics in the prometheus-compatible format, system components are already doing that; no other ways of exporting metrics is supported like using the OTLP push based protocol

Reviewer: ```suggestion
  - For instrumentation, a workload must expose metrics in the prometheus-compatible format. System components are already doing that. No other way of exporting metrics, like using the OTLP push-based protocol,  is supported.
```

  - Collection of metrics from system and custom workloads is done by a lightweight prometheus installation; configuration of collection can be defined at runtime but the storage might not scale in accordingly requiring adjustments at deploy time; configuration of outputs is not possible at runtime and only prometheus-specific protocols are supported (forward or federation)
  - Prometheus is collecting and storing the metrics; setup is non-scalable and resource settings cannot be configured at runtime
  - Reporting is done via a Grafana installation which loads pre-bundled dashboards; dashboards can be added at runtime
### Traces
  - Traces needs to be propagated via the Zipkin B3 protocol, which is supported by the istio infrastructure. From workload perspective, one needs to propagate that traces with requests and then can send additional span data to the JAeger collector in the Jaeger or zipkin protocol. The Istio, Serverless and Eventing component are supporting that already.

>   - The tracing component is based on the Jaeger all-in-one deployment which is acting as collector, not being scalable independent from the backend part. There are well-defined services in the cluster to which an application can push the span data to. However, it is tightly coupled to the related backend and does not provide any customiation, espacially no integration into other systems.

Reviewer: ```suggestion
  - The Tracing component is based on the Jaeger all-in-one deployment, which acts as collector and is not scalable independently from the backend part. There are well-defined services in the cluster to which an application can push the span data. However, it is tightly coupled to the related backend and does not provide any customization, especially no integration into other systems.
```

>   - The storage is based on a restricted in-memory store. So it is non-scalable and data will be lost on restarts.

Reviewer: ```suggestion
  - The storage is based on a restricted in-memory store. Thus, it is non-scalable and data is lost on restarts.
```

>   - The Visualization can be done via the explore tab in Grafana or the bundled Jaeger UI. Again, no further customization is possible here.

Reviewer: ```suggestion
  - The visualization can be done with the "Explore" tab in Grafana or the bundled Jaeger UI. Again, no further customization is possible here.
```

### Drawbacks

> At a first glance, the current solution provides a feature-rich end-to-end setup. However, at a second look, users notice major drawbacks and usually need additional stacks.

Reviewer: ```suggestion
At a first glance, the current solution provides a feature-rich end-to-end setup. However, at a second glance, users notice major drawbacks and usually need additional stacks.
```

## Architecture

> A new preconfigured agent layer gets introduced which is responsible for collecting all telemetry data without being bound to any backend. Hereby, the collection is very dependent on the signal type. The agents can be configured at runtime with different signal pipelines with basic filtering (inclusion and exclusion of signals) and outputs, so that the agents start shipping the signals through the pipelines to the configured backends. The dynamic configuration and management of the agent is handled by a new operator, which is configured using Kubernetes resources. The agents and the new operator are bundled in a new core component called `telemetry`. The existing Kyma backends and UIs will be just one possible solution to integrate with. The users will still be able to install them manually with a blueprint, but they will no longer be part of the Kyma offering.

Reviewer: ```suggestion
The proposal introduces a new preconfigured agent layer that's not bound to any backend. The agent layer is responsible for collecting all telemetry data, depending on the signal type.

Users can configure the agents at runtime with different signal pipelines using basic filtering (inclusion and exclusion of signals) and outputs, so that the agents start shipping the signals through the pipelines to the configured backends. The dynamic configuration and management of the collector is handled by a new operator, which is configured using Kubernetes resources. The agents and the new operator are bundled in a new core component called `telemetry`.

The existing Kyma backends and UIs will be just one possible solution to integrate with. The users will still be able to install them manually with a blueprint, but they will no longer be part of the Kyma offering.
```

![b](./assets/strategy-future.drawio.svg)

> As mentioned before, the technology and protocols for the signal collection are different for the signal types: Logs are tailed from container log files, metrics usually are pulled using the Prometheus format, and traces are pushed with OTLP. With that, also the pre-integration (so that typical signals are collected instantly) is different per type.

Reviewer: ```suggestion
As mentioned before, the technology and protocols for the signal collection depend on the signal types: Logs are tailed from container log files, metrics usually are pulled using the Prometheus format, and traces are pushed with OTLP. With that, also the pre-integration (so that typical signals are collected instantly) is different per type.
```
